master/buildbot/reporters/bitbucketserver.py

```diff
@@ -125,6 +129,201 @@ def send(self, build):
+        if not base_url:
+            config.error("Parameter base_url has to be given")
+        if token is not None and auth is not None:
+            config.error("Only one authentication method can be given "
+                "(token or auth)")
+        super().checkConfig(**kwargs)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, base_url, token=None, auth=None,
+                        name=None, nameSuffix=None, startDescription=None,
+                        endDescription=None, key=None, parent=None,
+                        buildNumber=None, ref=None, duration=None,
+                        testResults=None, verbose=False, debug=None,
+                        verify=None, **kwargs):
+        yield super().reconfigService(**kwargs)
+        self._name = name
+        self.nameSuffix = nameSuffix
+        self.startDescription = startDescription or 'Build started.'
+        self.endDescription = endDescription or 'Build done.'
+        self.key = key or Interpolate('%(prop:buildername)s')
+        self._parent = parent
+        self.buildNumber = buildNumber or Interpolate('%(prop:buildnumber)s')
+        self.ref = ref
+        self.duration = duration
+
+        if testResults:
+            self.testResults = testResults
+        else:
+            @util.renderer
+            def r_testresults(props):
+                failed = props.getProperty("tests_failed", 0)
+                skipped = props.getProperty("tests_skipped", 0)
+                successful = props.getProperty("tests_successful", 0)
+                if any([failed, skipped, successful]):
+                    return {
+                        "failed": failed,
+                        "skipped": skipped,
+                        "successful": successful
+                    }
+                return None
+            self.testResults = r_testresults
+
+        self.verbose = verbose
+        headers = {}
+        if token:
+            headers["Authorization"] = "Bearer {}".format(token)
+        self._http = yield httpclientservice.HTTPClientService.getService(
+            self.master, base_url, auth=auth, headers=headers, debug=debug,
+            verify=verify)
+
+    def createStatus(self, projKey, repoSlug, sha, state, url, key, parent,
+                     buildNumber, ref, description, name, duration,
+                     testResults):
+        payload = {
+            'state': state,
+            'url': url,
+            'key': key,
+            'parent': parent,
+            'ref': ref,
+            'buildNumber': buildNumber,
+            'description': description,
+            'name': name,
+            'duration': duration,
+            'testResults': testResults
+        }
+
+        if self.verbose:
+            log.msg("Sending payload: '{}' for {}/{} {}.".format(payload,
+                projKey, repoSlug, sha))
+
+        return self._http.post(STATUS_CORE_API_URL.format(projKey=projKey,
```

Comment:
Wondering if extracting the url computation into separate line would be more readable.